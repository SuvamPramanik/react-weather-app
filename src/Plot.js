/* global Plotly */
import PropTypes from 'prop-types';
import React, {Component} from 'react';

class Plot extends Component {

  constructor () {
    super();
    this.renderPlot = this.renderPlot.bind(this);
  }

  shouldComponentUpdate (nextProps) {
    return !this.props.xData.equals(nextProps.xData) || !this.props.yData.equals(nextProps.yData);
  }

  renderPlot () {
    Plotly.newPlot('plot', [{
      x: this.props.xData.toJS(),
      y: this.props.yData.toJS(),
      type: this.props.type
    }], {
      margin: {
        t: 50, r: 0, l: 30
      },
      xaxis: {
        gridcolor: 'transparent'
      }
    }, {
      displayModeBar: false
    });
    document.getElementById('plot').on('plotly_click', this.props.onPlotClick);
  }

  componentDidMount () {
    this.renderPlot();
  }

  componentDidUpdate () {
    this.renderPlot();
  }

  render () {
    return (
      <div id='plot' />
    );
  }
}

Plot.propTypes = {
  xData: PropTypes.array,
  yData: PropTypes.array,
  type: PropTypes.string,
  onPlotClick: PropTypes.func
};
export default Plot;
