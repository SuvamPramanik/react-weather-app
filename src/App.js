import PropTypes from 'prop-types';
import React, {Component} from 'react';
import xhr from 'xhr';
import './App.css';
import ErrorPage from './ErrorPage.js';
import Plot from './Plot.js';
import {changeLocation, setData, setDates, setErrorStatus, setSelectedDate, setSelectedTemp, setTemps} from './action.js';
import {connect} from 'react-redux';

const API_KEY = '50ebb139a30adf57eb66817bf692b98e';

class App extends Component {
  constructor () {
    super();

    this.fetchWeatherData = this.fetchWeatherData.bind(this);
    this.updateLocation = this.updateLocation.bind(this);
    this.onPlotClick = this.onPlotClick.bind(this);
  }

  fetchWeatherData (evt) {
    evt.preventDefault();
    const location = encodeURIComponent(this.props.location);
    if (location !== '') {
      const apiPrefix = 'https://api.openweathermap.org/data/2.5/forecast?q=';
      const apiSuffix = '&APPID=' + API_KEY + '&units=metric';
      const api = apiPrefix + location + apiSuffix;

      // Used for getting the reference inside other components
      const self = this;

      xhr({
        url: api
      }, function (err, data) {
        if (data.statusCode === 404) {
          self.props.dispatch(setErrorStatus(true));
        } else {
          var body = JSON.parse(data.body);
          var dates = [];
          var temps = [];
          for (var item of body.list) {
            dates.push(item.dt_txt);
            temps.push(item.main.temp);
          }
          self.props.dispatch(setErrorStatus(false));
          self.props.dispatch(setData(body));
          self.props.dispatch(setDates(dates));
          self.props.dispatch(setTemps(temps));
          self.props.dispatch(setSelectedTemp(null));
          self.props.dispatch(setSelectedDate(''));
        }
      });
    }
  }

  updateLocation (evt) {
    this.props.dispatch(changeLocation(evt.target.value));
  }

  onPlotClick (data) {
    if (data.points) {
      this.props.dispatch(setSelectedDate(data.points[0].x));
      this.props.dispatch(setSelectedTemp(data.points[0].y));
    }
  }

  render () {
    let currTemp = 'Loading weather data....';
    if (this.props.data.list) {
      currTemp = this.props.data.list[0].main.temp;
    }
    return (
      <div className={'weather-div'}>
        <h1>Weather Page</h1>
        <form onSubmit={this.fetchWeatherData}>
          <label>I want to know the weather for
            <input
              id={'place-input'}
              placeholder={'City, Country'}
              type='text'
              value={this.props.location}
              onChange={this.updateLocation} />
          </label>
        </form>
        {/*
          Render the current temperature and the forecast if we have data
          otherwise return null
        */}
        { (this.props.fetchError) ? (
          <ErrorPage />
        ) : (this.props.data.list) ? (
          <div className='wrapper'>
            <p className='temp-wrapper'>
              <span className='temp'>
                {this.props.tempSelected ? this.props.tempSelected : currTemp}
              </span>
              <span className='temp-symbol'>Â°C</span>
              <span className='temp-date'>
                { this.props.tempSelected ? this.props.dateSelected : ''}
              </span>
            </p>
            <h2>Forecast</h2>
            <Plot
              xData={this.props.dates}
              yData={this.props.temps}
              type='scatter'
              onPlotClick={this.onPlotClick} />
          </div>) : null }
      </div>
    );
  }
}

function mapStateToProps (state) {
  return state;
}

App.propTypes = {
  location: PropTypes.string,
  dispatch: PropTypes.func,
  data: PropTypes.object,
  fetchError: PropTypes.bool,
  tempSelected: PropTypes.number,
  dateSelected: PropTypes.string,
  dates: PropTypes.array,
  temps: PropTypes.array
};

export default connect(mapStateToProps)(App);
