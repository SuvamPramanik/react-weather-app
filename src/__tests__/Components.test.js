import Plot from '../Plot';
import React from 'react';
import renderer from 'react-test-renderer';
import {App} from '../App';
import {fromJS} from 'immutable';

describe('Components testing', () => {
  describe('<App />', () => {
    it('renders correctly', () => {
      const tree = renderer.create(<App redux={fromJS({})}/>).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  describe('<Plot />', function () {
    global.Plotly = {
      newPlot: () => {}
    };
    global.document = {
      getElementById: function () {
        return {
          on: () => {}
        };
      }
    };

    it('renders correctly', function () {
      const tree = renderer.create(<Plot xData={fromJS({})} yData={fromJS({})} />).toJSON();
      expect(tree).toMatchSnapshot();
    });
  });
});
