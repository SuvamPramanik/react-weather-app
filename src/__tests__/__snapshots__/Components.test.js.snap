// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components testing <App /> renders correctly 1`] = `
<div
  className="weather-div"
>
  <h1>
    Weather Page
  </h1>
  <form
    onSubmit={[Function]}
  >
    <label>
      I want to know the weather for
      <input
        id="place-input"
        onChange={[Function]}
        placeholder="City, Country"
        type="text"
        value={undefined}
      />
    </label>
  </form>
</div>
`;
